engines/sherlock/resources.cpp

```diff
@@ -0,0 +1,471 @@
+
+	load(name, f);
+
+	f.close();
+}
+
+/**
+ * Load a cache entry based on a passed stream
+ */
+void Cache::load(const Common::String &name, Common::SeekableReadStream &stream) {
+	// First check if the entry already exists
+	if (_resources.contains(name))
+		return;
+
+	int32 signature = stream.readUint32BE();
+	stream.seek(0);
+
+	// Allocate a new cache entry
+	_resources[name] = CacheEntry();
+	CacheEntry &cacheEntry = _resources[name];
+
+	// Check whether the file is compressed
+	if (signature == MKTAG('L', 'Z', 'V', 26)) {
+		// It's compressed, so decompress the file and store it's data in the cache entry
+		Common::SeekableReadStream *decompressed = _vm->_res->decompressLZ(stream);
+		cacheEntry.resize(decompressed->size());
+		decompressed->read(&cacheEntry[0], decompressed->size());
+
+		delete decompressed;
+	} else {
+		// It's not, so read the raw data of the file into the cache entry
+		cacheEntry.resize(stream.size());
+		stream.read(&cacheEntry[0], stream.size());
+	}
+}
+
+/**
+ * Get a file from the cache
+ */
+Common::SeekableReadStream *Cache::get(const Common::String &filename) const {
+	// Return a memory stream that encapsulates the data
+	const CacheEntry &cacheEntry = _resources[filename];
+	return new Common::MemoryReadStream(&cacheEntry[0], cacheEntry.size());
+}
+
+/*----------------------------------------------------------------*/
+
+Resources::Resources(SherlockEngine *vm): _vm(vm), _cache(vm) {
+	_resourceIndex = -1;
+
+	addToCache("vgs.lib");
+	addToCache("talk.lib");
+	addToCache("sequence.txt");
+	addToCache("journal.txt");
+	addToCache("portrait.lib");
+}
+
+/**
+ * Adds the specified file to the cache. If it's a library file, takes care of
+ * loading it's index for future use
+ */
+void Resources::addToCache(const Common::String &filename) {
+	_cache.load(filename);
+
+	// Check to see if the file is a library
+	Common::SeekableReadStream *stream = load(filename);
+	uint32 header = stream->readUint32BE();
+	if (header == MKTAG('L', 'I', 'B', 26))
+		loadLibraryIndex(filename, stream);
+
+	delete stream;
+}
+
+/**
+ * Adds a resource from a library file to the cache
+ */
+void Resources::addToCache(const Common::String &filename, const Common::String &libFilename) {
+	// Get the resource
+	Common::SeekableReadStream *stream = load(filename, libFilename);
+
+	_cache.load(filename, *stream);
+
+	delete stream;
+}
+
+/**
+ * Adds a given stream to the cache under the given name
+ */
+void Resources::addToCache(const Common::String &filename, Common::SeekableReadStream &stream) {
+	_cache.load(filename, stream);
+}
+
+/**
+ * Returns a stream for a given file
+ */
+Common::SeekableReadStream *Resources::load(const Common::String &filename) {
+	// First check if the file is directly in the cache
+	if (_cache.isCached(filename))
+		return _cache.get(filename);
+
+	// Secondly, iterate through any loaded library file looking for a resource
+	// that has the same name
+	LibraryIndexes::iterator i;
```

Comment:
It looks like this should rather be move inside the for loop, the variable `i` never seems to be used outside of it.